archive/imo/imo2008_q4.lean

```diff
@@ -0,0 +1,135 @@
+/-
+Copyright (c) 2021 Manuel Candales. All rights reserved.
+Released under Apache 2.0 license as described in the file LICENSE.
+Authors: Manuel Candales
+-/
+import data.real.basic
+import data.real.sqrt
+
+/-!
+# IMO 2008 Q4
+Find all functions `f : (0,∞) → (0,∞)` (so, `f` is a function from the positive real
+numbers to the positive real numbers) such that
+      ```
+      (f(w)^2 + f(x)^2)/(f(y^2) + f(z^2)) = (w^2 + x^2)/(y^2 + z^2)
+      ```
+for all positive real numbers `w`, `x`, `y`, `z`, satisfying `wx = yz`.
+
+# Solution
+The desired theorem is that either `f = λ x, x` or `f = λ x, 1/x`
+-/
+
+open real
+
+theorem imo2008_q4
+  (f : ℝ → ℝ)
+  (H₁ : ∀ x > 0, f(x) > 0) :
+  (∀ w x y z : ℝ, 0 < w → 0 < x → 0 < y → 0 < z → w*x = y*z →
+    (f(w)^2 + f(x)^2)/(f(y^2) + f(z^2)) = (w^2 + x^2)/(y^2 + z^2)) ↔
+  ((∀ x > 0, f(x) = x) ∨ (∀ x > 0, f(x) = 1/x)) :=
+begin
+  split, swap,
+  -- proof that f(x) = x and f(x) = 1/x satisfy the condition
+  { rintros (h | h),
+    { intros w x y z hw hx hy hz hprod,
+      rw [(h w hw), (h x hx), (h (y^2) (pow_pos hy 2)), (h (z^2) (pow_pos hz 2))] },
+    { intros w x y z hw hx hy hz hprod,
+      rw [(h w hw), (h x hx), (h (y^2) (pow_pos hy 2)), (h (z^2) (pow_pos hz 2))],
+      have hy2z2 : y^2 + z^2 ≠ 0 := ne_of_gt (add_pos (pow_pos hy 2) (pow_pos hz 2)),
+      have hz2y2 : z^2 + y^2 ≠ 0 := ne_of_gt (add_pos (pow_pos hz 2) (pow_pos hy 2)),
+      have hp2 : w^2 * x^2 = y^2 * z^2, { rw [(mul_pow w x 2).symm, (mul_pow y z 2).symm, hprod] },
+      field_simp [(ne_of_gt hw), (ne_of_gt hx), (ne_of_gt hy), (ne_of_gt hz), hy2z2, hz2y2, hp2],
+      ring } },
+
+  -- proof that the only solutions are f(x) = x or f(x) = 1/x
+  intro H₂,
+  have h₀ : f(1) ≠ 0, { specialize H₁ 1 zero_lt_one, exact ne_of_gt H₁ },
+
+  have h₁ : f(1) = 1,
+  { specialize H₂ 1 1 1 1 zero_lt_one zero_lt_one zero_lt_one zero_lt_one rfl,
+    norm_num at H₂,
+    rw [(two_mul (f(1)^2)).symm, (two_mul (f 1)).symm] at H₂,
+    rw mul_div_mul_left (f(1)^2) (f 1) two_ne_zero at H₂,
+    rwa ← (div_eq_iff h₀).mpr (pow_two (f 1)) },
+
+  have h₂ : ∀ x > 0, (f x - x) * (f x - 1/x) = 0,
+  { intros x hx,
+    have h1xss : 1 * x = (sqrt x) * (sqrt x), { rw [one_mul, mul_self_sqrt (le_of_lt hx)] },
+    specialize H₂ 1 x (sqrt x) (sqrt x) zero_lt_one hx (sqrt_pos.mpr hx) (sqrt_pos.mpr hx) h1xss,
+    rw [h₁, (one_pow 2), (sqr_sqrt (le_of_lt hx)), (two_mul (f x)).symm, (two_mul x).symm] at H₂,
```

Comment:
You can replace rewrite rules of the form `h.symm` by `← h` in `rw`/`simp`/`simp_rw`/... (in various places in your proof)